wallet/datastore.go

```diff
@@ -431,47 +354,419 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
+		return fmt.Errorf("failed to create tx for wallet linking: %w", err)
+	}
+	// add tx to ctx for future
+	ctx = context.WithValue(ctx, appctx.DatabaseTransactionCTXKey, tx)
+	defer pg.RollbackTx(tx)
+
+	id, err := uuid.FromString(ID)
+	if err != nil {
+		return errorutils.Wrap(err, "invalid id")
+	}
+
+	// connect custodian link (does the link limit checking in insert)
+	if err = pg.ConnectCustodialWallet(ctx, &CustodianLink{
+		WalletID:  &id,
+		Custodian: depositProvider,
+		LinkingID: &providerLinkingID,
+	}, userDepositDestination); err != nil {
+		sublogger.Error().Err(err).
+			Msg("failed to insert new custodian link")
+		return fmt.Errorf("failed to insert new custodian link: %w", err)
+	}
+
+	err = tx.Commit()
 	if err != nil {
 		return err
 	}
-	defer pg.RollbackTx(tx)
+	return nil
+}
+
+// CustodianLink - representation of wallet_custodian record
+type CustodianLink struct {
+	WalletID           *uuid.UUID   `json:"wallet_id" db:"wallet_id" valid:"uuidv4"`
+	Custodian          string       `json:"custodian" db:"custodian" valid:"in(uphold,brave,gemini,bitflyer)"`
+	CreatedAt          time.Time    `json:"created_at" db:"created_at" valid:"-"`
+	LinkedAt           time.Time    `json:"linked_at" db:"linked_at" valid:"-"`
+	DisconnectedAt     sql.NullTime `json:"disconnected_at" db:"disconnected_at" valid:"-"`
+	DepositDestination string       `json:"deposit_destination" db:"deposit_destination" valid:"-"`
+	LinkingID          *uuid.UUID   `json:"linking_id" db:"linking_id" valid:"uuid"`
+}
+
+// GetWalletIDString - get string version of the WalletID
+func (cl *CustodianLink) GetWalletIDString() string {
+	if cl.WalletID != nil {
+		return cl.WalletID.String()
+	}
+	return ""
+}
+
+// GetLinkingIDString - get string version of the LinkingID
+func (cl *CustodianLink) GetLinkingIDString() string {
+	if cl.LinkingID != nil {
+		return cl.LinkingID.String()
+	}
+	return ""
+}
+
+// GetCustodianLinkCount - get the wallet custodian link count across all wallets
+func (pg *Postgres) GetCustodianLinkCount(ctx context.Context, linkingID uuid.UUID) (int, int, error) {
+	// the count of linked wallets
+	used := 0
+	max := 0
+	var err error
 
-	maxLinkings, err := txGetMaxLinkingSlots(ctx, tx, providerLinkingID.String())
+	// create a sublogger
+	sublogger := logger(ctx).With().
+		Str("linking_id", linkingID.String()).
+		Logger()
+
+	sublogger.Debug().
+		Msg("starting GetCustodianLinkCount")
+
+	// get tx
+	_, tx, rollback, commit, err := getTx(ctx, pg)
 	if err != nil {
-		return errorutils.Wrap(err, "error looking up max linkings for wallet")
+		return 0, 0, fmt.Errorf("failed to create db transaction GetCustodianLinkByWalletID: %w", err)
 	}
+	// will rollback if tx created at this scope
+	defer rollback()
 
-	usedLinkings, err := txGetUsedLinkingSlots(ctx, tx, providerLinkingID.String())
+	// get used
+	stmt := `
+		select count(distinct(wallet_id)) as used from wallet_custodian where linking_id = $1
```

Comment:
should we update the logic in `txGetUsedLinkingSlots` to use this query? ( and call `txGetUsedLinkingSlots` )

Reply:
:+1:

Reply:
resolved: https://github.com/brave-intl/bat-go/pull/793/commits/dcbcd7ca13f72be4249e0760d01398e454c5bc66